Document behavior of endpoints with the feature EndpointSliceTerminatingCondition (#36791)
kubernetes/website: content/en/docs/concepts/workloads/pods/pod-lifecycle.md

      The containers in the Pod receive the TERM signal at different times and in an arbitrary
      order. If the order of shutdowns matters, consider using a `preStop` hook to synchronize.
      {{< /note >}}

> {{< feature-state for_k8s_version="v1.22" state="beta" >}}

Reviewer: I'm not sure if this feature state tag is appropriate here because this section also discusses behaviors that existed prior to the `EndpointSliceTerminatingCondition` feature.

Author: agree, removed

1. At the same time as the kubelet is starting graceful shutdown, the control plane evaluates whether
   shutting-down Pod needs to be removed from Endpoints (and, if enabled, EndpointSlice) objects where these represent
   a {{< glossary_tooltip term_id="service" text="Service" >}} with a configured
   {{< glossary_tooltip text="selector" term_id="selector" >}}.
   {{< glossary_tooltip text="ReplicaSets" term_id="replica-set" >}} and other workload resources
   no longer treat the shutting-down Pod as a valid, in-service replica. Pods that shut down slowly
   should not continue to serve traffic as normal and should be terminating connection. Without feature
   gate `EndpointSliceTerminatingCondition` (in beta starting Kubernetes version 1.22), load balancers (like the service proxy) remove the Pod from
   the list of endpoints as soon as the termination grace period _begins_. With the feature
   gate `EndpointSliceTerminatingCondition`, Endpoints are not removed for the terminating pods

>    and new status indication terminating state is exposed from Endpoint API.

Reviewer: Just so it's not confused with old Endpoint API, I would replace all Endpoints with EndpointSlice here. `Endpoint` API does not (and will not) support inclusion of terminating endpoints.

Reviewer: A reference to the [EndpointSlice conditions documentation]( https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#conditions) here would be helpful I think, because for backwards compatibly a terminating endpoint is always Ready=True, so if you care about the actual readiness of a terminating pod, you need to check a new condition `serving`.

Author: Updated

      The containers in the Pod receive the TERM signal at different times and in an arbitrary
      order. If the order of shutdowns matters, consider using a `preStop` hook to synchronize.
      {{< /note >}}

1. At the same time as the kubelet is starting graceful shutdown, the control plane evaluates whether

>    shutting-down Pod needs to be removed from Endpoints (and, if enabled, EndpointSlice) objects where these represent

Reviewer 2: Hi @SergeyKanzhelev .
Is a 'shutting-down' Pod the same as a terminating Pod?

Possible edit:
At the same time as the kubelet is starting graceful shutdown, the control plane evaluates whether a terminating Pod needs to be removed from any Endpoints and if enabled, EndpointSlice objects. 

Possible edit:
Pods that shut down slowly should not continue to serve traffic and instead terminate the connection.

Author: applied first and mentioned "regular traffic" in second edit to match the last sentence about load balancers directing regular traffic. I hope it clarifies things

      The containers in the Pod receive the TERM signal at different times and in an arbitrary
      order. If the order of shutdowns matters, consider using a `preStop` hook to synchronize.
      {{< /note >}}

1. At the same time as the kubelet is starting graceful shutdown, the control plane evaluates

>    whether a terminating Pod needs to be removed from any Endpoints and if enabled,

Reviewer 3: ```suggestion
   whether a terminating Pod needs to be removed from any Endpoints and
```

>    EndpointSlice object which represent

Reviewer 3: ```suggestoin
1. At the same time as the kubelet is starting graceful shutdown, the control plane evaluates
   whether to remove that shutting-down Pod from EndpointSlice (and Endpoints) objects, where
   those objects represent
```

   a {{< glossary_tooltip term_id="service" text="Service" >}} with a configured
   {{< glossary_tooltip text="selector" term_id="selector" >}}.
   {{< glossary_tooltip text="ReplicaSets" term_id="replica-set" >}} and other workload resources
   no longer treat the shutting-down Pod as a valid, in-service replica. Pods that shut down slowly
   should not continue to serve regular traffic and should be terminating connection.

>    Some pods require the graceful termination - for example session draining and completion. Without the feature
>    gate `EndpointSliceTerminatingCondition` (in beta starting Kubernetes version `1.22`),

Reviewer 3: ```suggestion
   Some pods work best when you use graceful termination - for example, you might want that
   for graceful session draining and completion. If you don't have the the
   `EndpointSliceTerminatingCondition` feature gate enabled in your cluster, then
```

Reviewer 4: This maybe a nit. However, do we want to use graceful twice in the first sentence? Session draining & completion, afaik, should be graceful?

   load balancers (like the service proxy) remove the Pod from

>    the list of Endpoints in EndpointSlices as soon as the termination grace period _begins_. With the feature

Reviewer 3: ```suggestion
   the list of endpoints in EndpointSlices as soon as the termination grace period _begins_. With the feature
```

   the list of EndpointSlices as soon as the termination grace period _begins_. With the feature

>    gate `EndpointSliceTerminatingCondition`, EndpointSlices are not removed for the terminating pods

Reviewer 3: the list of EndpointSlices as soon as the termination grace period _begins_.

   If your cluster has the default behavior (the EndpointSliceTerminatingCondition
   feature gate is enabled), then EndpointSlices continue to include Pods even when
   those Pods start to terminate,
```

   and a status indicating [terminating state](/docs/concepts/services-networking/endpoint-slices/#conditions)

>    is exposed from Endpoint API. Terminating EndpointSlice always have `ready` status

Reviewer 3: This doesn't make sense. For a start, there's no Endpoint resource kind in Kubernetes (the precursor to EndpointSlice was “Endpoints”, plural. But also, the explanation looks all mixed up.

Author: yes, I fixed it now. I was trying not to touch it, but ended up adding an example section to explain this. It's easier to understand with the example

>    is exposed from Endpoint API. Terminating Endpoints always have `ready` status

Reviewer 3: ```suggestion
   is exposed from the EndpointSlice API (and the legacy Endpoints API). Terminating
  endpoints always have their `ready` status
```

   load balancers (like the service proxy) remove the Pod from
   the list of Endpoints in EndpointSlices as soon as the termination grace period _begins_. With the feature
   gate `EndpointSliceTerminatingCondition`, Endpoints are not removed from EndpointSlices for the terminating pods
   and a status indicating [terminating state](/docs/concepts/services-networking/endpoint-slices/#conditions)
   is exposed from Endpoint API. Terminating Endpoints always have `ready` status
   as `false` for backward compatibility, so load balancers will not use it for regular traffic.
   If traffic draining on terminating pod is needed, the actual readiness can be checked as a condition `serving`.
1. When the grace period expires, the kubelet triggers forcible shutdown. The container runtime sends
   `SIGKILL` to any processes still running in any container in the Pod.
   The kubelet also cleans up a hidden `pause` container if that container runtime uses one.
1. The kubelet triggers forcible removal of Pod object from the API server, by setting grace period
   to 0 (immediate deletion).
1. The API server deletes the Pod's API object, which is then no longer visible from any client.

> #### Example flow with endpoint termination

Reviewer 5: I appreciate the detailed example for explaining the technical points.
However, my feeling is that this section doesn't belong here as a concept.
The concept page is already pretty long.
Maybe we want to find a new home (e.g. a task page) for this example?

Reviewer 3: Could we move this into a new tutorial that follows on from [Connecting Applications With Services](https://kubernetes.io/docs/tutorials/services/connect-applications-service/)?

The following is the example of the flow described in previous section.

Let's say you have a Deployment containing of a single `nginx` replica
(just for demonstration purposes) and a Service:

> ``` yaml

Reviewer 5: looks like we don't need the extra space before 'yaml'.
Similar issues are found below.

Reviewer 3: Could we move this into a new tutorial that follows on from [Connecting Applications With Services](https://kubernetes.io/docs/tutorials/services/connect-applications-service/)?

apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      terminationGracePeriodSeconds: 120
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
        lifecycle:
          preStop:
            exec:
              # just hang around for the duration of terminationGracePeriodSeconds
              command: [
                "/bin/sh", "-c", "sleep 120"
              ]

---

apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  selector:
    app: nginx
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
```

Once the Pod and Service are running, you can get the name of EndpointSlice:

``` sh
kubectl get endpointslice

NAME                  ADDRESSTYPE   PORTS   ENDPOINTS                 AGE
nginx-service-6tjbr   IPv4          80      10.12.1.199,10.12.1.201   22m
```

You can see it's status and validate that there is one Endpoint registered:

``` sh

> > kubectl get endpointslice -o json nginx-service-6tjbr

Reviewer 3: Please [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output) (and also, don't include a prompt)

> ``` sh
> > kubectl get endpointslice -o json nginx-service-6tjbr
> {
>     "addressType": "IPv4",
>     "apiVersion": "discovery.k8s.io/v1",
>     "endpoints": [
>         {
>             "addresses": [
>                 "10.12.1.201"
>             ],
>             "conditions": {
>                 "ready": true,
>                 "serving": true,
>                 "terminating": false
> ```

Reviewer 5: ````suggestion
```shell
kubectl get endpointslice -o json nginx-service-6tjbr
```

The output is similar to this:

```
{
    "addressType": "IPv4",
    "apiVersion": "discovery.k8s.io/v1",
    "endpoints": [
        {
            "addresses": [
                "10.12.1.201"
            ],
            "conditions": {
                "ready": true,
                "serving": true,
                "terminating": false
```
````

Now let's terminate the pod and validate that the pod is being terminated
respecting the graceful termination period configuration:

``` sh
kubectl delete pod nginx-deployment-7768647bf9-b4b9s
```

All pods:

```
kubectl get pods

> 

Reviewer 3: Please [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output)

NAME                                READY   STATUS        RESTARTS      AGE
nginx-deployment-7768647bf9-b4b9s   1/1     Terminating   0             4m1s
nginx-deployment-7768647bf9-rkxlw   1/1     Running       0             8s
```

You can see that the new pod got scheduled.

With the feature gate `EndpointSliceTerminatingCondition` enabled by default
while the new Endpoint is being created for the new Pod, the old endpoint is
still around in the terminating state:

```
kubectl get endpointslice -o json nginx-service-6tjbr

> 

Reviewer 3: Please [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output)

      The containers in the Pod receive the TERM signal at different times and in an arbitrary
      order. If the order of shutdowns matters, consider using a `preStop` hook to synchronize.
      {{< /note >}}
1. At the same time as the kubelet is starting graceful shutdown, the control plane evaluates whether to remove that shutting-down Pod from EndpointSlice (and Endpoints) objects, where those objects represent
   a {{< glossary_tooltip term_id="service" text="Service" >}} with a configured
   {{< glossary_tooltip text="selector" term_id="selector" >}}.
   {{< glossary_tooltip text="ReplicaSets" term_id="replica-set" >}} and other workload resources
   no longer treat the shutting-down Pod as a valid, in-service replica. Pods that shut down slowly
   should not continue to serve regular traffic and should be terminating connection.
   Some pods work best when you use graceful termination - for example, you might want that
   for session draining and completion. If you don't have the the
   `EndpointSliceTerminatingCondition` feature gate enabled in your cluster, then

>    load balancers (like the service proxy) remove the Pod from

Author: I don't think it is talking about kube-proxy. Pods connectivity suppose to stay active. I believe it refers to the proxies for Services. I kept the original wording

      The containers in the Pod receive the TERM signal at different times and in an arbitrary
      order. If the order of shutdowns matters, consider using a `preStop` hook to synchronize.
      {{< /note >}}

> 1. At the same time as the kubelet is starting graceful shutdown, the control plane evaluates whether to remove that shutting-down Pod from EndpointSlice (and Endpoints) objects, where those objects represent

Reviewer 6: ```suggestion
1. At the same time as the kubelet is starting graceful shutdown of the Pod, the control plane evaluates whether to remove that shutting-down Pod from EndpointSlice (and Endpoints) objects, where those objects represent
```

I read it as if was the kubelet that started to shutdown

   a {{< glossary_tooltip term_id="service" text="Service" >}} with a configured
   {{< glossary_tooltip text="selector" term_id="selector" >}}.
   {{< glossary_tooltip text="ReplicaSets" term_id="replica-set" >}} and other workload resources
   no longer treat the shutting-down Pod as a valid, in-service replica. Pods that shut down slowly
   should not continue to serve regular traffic and should start terminating and finish processing open connections.
   Some applications need to go beyond finishing open connections and need more graceful termination -

>    for example session draining and completion. Endpoints that represent the terminating pods

Reviewer 3: ```suggestion
   for example: session draining and completion. Any endpoints that represent the terminating pods
```

_The extra “Any” lets us write “endpoints” all-lowercase, to reduce confusion with the API kind named Endpoints._

   are not immediately removed from EndpointSlices,
   and a status indicating [terminating state](/docs/concepts/services-networking/endpoint-slices/#conditions)
   is exposed from the EndpointSlice API (and the legacy Endpoints API). Terminating
   endpoints always have their `ready` status
   as `false` (for backward compatibility with versions before 1.26),
   so load balancers will not use it for regular traffic.
   If traffic draining on terminating pod is needed, the actual readiness can be checked as a condition `serving`.
   You can find more details on how to implement connections draining
   in the tutorial [Pods And Endpoints Termination Flow](/docs/tutorials/services/pods-and-endpoint-termination-flow/)

{{<note>}}
If you don't have the `EndpointSliceTerminatingCondition` feature gate enabled

> in your cluster (before the version 1.26), then load balancers (like the service proxy)

Reviewer 3: in your cluster (the gate is on by default from Kubernetes 1.26), then the Kubernetes control plane

> in your cluster (the gate is on by default from Kubernetes 1.26), then the Kubernetes control

Reviewer 6: it was beta in 1.22 https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

Reviewer 6: last supported release is 1.24 https://kubernetes.io/releases/

Reviewer 3: We should document this detail until v1.25 is out of support. v1.25 is the last release that supported setting `EndpointSliceTerminatingCondition` false.

Author: fixed

> in your cluster (before the version 1.26), then load balancers (like the service proxy)
> remove the Pod from the list of endpoints in EndpointSlices as soon as the

Reviewer 3: ```suggestion
in your cluster (the gate is on by default from Kubernetes 1.26), then the Kubernetes control
plane removes a Pod from any relevant EndpointSlices as soon as the Pod's
```
